MazerPlatformer/GameWorld.cs

```diff
@@ -145,61 +148,88 @@ public void Initialize()
+                                // pickup points
+                                result = DetermineNewLevelPoints(thePlayer, otherObject)
+                                    .Bind(levelPoints => thePlayer.UpdateComponentByType(Component.ComponentType.Points, levelPoints))
+                                    .Map(o => Nothing);
+                                break;
+                        }
+
+                        return result;
+                    });
+
+            Either<IFailure, int> DetermineNewHealth(GameObject gameObject, GameObject otherObject1)
             {
-                // deal damage
-                var hitPoints = otherObject.FindComponentByType(Component.ComponentType.HitPoints).Value;
-                var myHealth = thePlayer.FindComponentByType(Component.ComponentType.Health).Value;
-                var newHealth = (int)myHealth - (int)hitPoints;
-                thePlayer.UpdateComponentByType(Component.ComponentType.Health, newHealth);
-
-                if (newHealth <= 0)
-                {
-                    _level.PlayLoseSound();
-                    _playerDied = true;
-                    OnPlayerDied?.Invoke(thePlayer.Components);
-                }
+                return from hitPointsComponent in otherObject1.FindComponentByType(Component.ComponentType.HitPoints)
+                        .ToEither(NotFound.Create("Could not find hit-point component"))
+                    from healthComponent in gameObject.FindComponentByType(Component.ComponentType.Health)
+                        .ToEither(NotFound.Create("Could not find health component"))
+                    let myHealth = (int) healthComponent.Value
+                    let hitPoints = (int) hitPointsComponent.Value
+                    select myHealth - hitPoints;
             }
 
-            if (npcType == Npc.NpcTypes.Pickup)
+            Either<IFailure, int> DetermineNewLevelPoints(GameObject thePlayer1, GameObject gameObject1)
             {
-                // pickup points
-                var pickupPoints = (int)otherObject.FindComponentByType(Component.ComponentType.Points).Value;
-                var myPoints = (int)thePlayer.FindComponentByType(Component.ComponentType.Points).Value;
-                var levelPoints = myPoints + pickupPoints;
-                thePlayer.UpdateComponentByType(Component.ComponentType.Points, levelPoints);
+                return from pickupPointsComponent in gameObject1.FindComponentByType(Component.ComponentType.Points)
+                        .ToEither(NotFound.Create("Could not find hit-point component"))
+                    from myPointsComponent in thePlayer1.FindComponentByType(Component.ComponentType.Points)
+                        .ToEither(NotFound.Create("Could not find hit-point component"))
+                    let myPoints = (int) myPointsComponent.Value
+                    let pickupPoints = (int) pickupPointsComponent.Value
+                    select myPoints + pickupPoints;
```

Comment:
This is a good example of linq query syntax used in LanguageExt.
Makes declarative usage of logic.